includes/Checker/Checks/Abstract_File_Check.php

```diff
@@ -0,0 +1,219 @@
+				$files,
+				static function( $file ) use ( $lookup ) {
+					return isset( $lookup[ pathinfo( $file, PATHINFO_EXTENSION ) ] );
+				}
+			)
+		);
+	}
+
+	/**
+	 * Filters a given list of files to only contain those where the file name matches the given regular expression.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param array  $files List of absolute file paths.
+	 * @param string $regex Regular expression for file paths to match.
+	 * @return array Filtered $files list.
+	 */
+	final protected static function filter_files_by_regex( array $files, $regex ) {
+		return preg_grep( $regex, $files );
+	}
+
+	/**
+	 * Performs a regular expression match on the file contents of the given list of files.
+	 *
+	 * This is a wrapper around the native `preg_match()` function that will match the first occurrence within the
+	 * list of files.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $pattern The pattern to search for.
+	 * @param array  $files   List of absolute file paths.
+	 * @param array  $matches Optional. Array to store the matches, passed by reference. Similar to `preg_match()`,
+	 *                        `$matches[0]` will contain the text that matched the full pattern, `$matches[1]` will
+	 *                        have the text that matched the first captured parenthesized subpattern, and so on.
+	 * @return string|bool File path if a match was found, false otherwise.
+	 */
+	final protected static function file_preg_match( $pattern, array $files, array &$matches = null ) {
+		foreach ( $files as $file ) {
+			$contents = self::file_get_contents( $file );
+			if ( preg_match( $pattern, $contents, $m ) ) {
+				$matches = $m;
+				return $file;
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * Performs a check indicating if the needle is contained in the file contents of the given list of files.
+	 *
+	 * This is a wrapper around the native `str_contains()` function that will find the needle within the list of
+	 * files.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param array  $files  List of absolute file paths.
+	 * @param string $needle The substring to search for.
+	 * @return string|bool File path if needle was found, false otherwise.
+	 */
+	final protected static function file_str_contains( array $files, $needle ) {
+		foreach ( $files as $file ) {
+			$contents = self::file_get_contents( $file );
+			if ( str_contains( $contents, $needle ) ) {
+				return $file;
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * Gets the contents of the given file.
+	 *
+	 * This is effectively a caching wrapper around the native `file_get_contents()` function.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $file The file name.
+	 * @return string The file contents.
+	 */
+	private static function file_get_contents( $file ) {
+		if ( isset( self::$file_contents_cache[ $file ] ) ) {
+			return self::$file_contents_cache[ $file ];
+		}
+
+		self::$file_contents_cache[ $file ] = file_get_contents( $file );
+
+		return self::$file_contents_cache[ $file ];
+	}
```

Comment:
So we are using this as a cache?

Reply:
Yes, this is particularly relevant as the same file contents may be needed again across multiple checks in a single request. Note that this was inspired by https://github.com/WordPress/plugin-check/blob/trunk/checks/check-base.php#L159 which does the same thing, which makes sense to me.